File: ocaml/21-30/26.ml
(* Generate the combinations of K distinct objects chosen from the N elements of a list. (medium)
 *
 * In how many ways can a committee of 3 be chosen from a group of 12 people? We all know that there are C(12,3) = 220 possibilities (C(N,K) denotes the well-known binomial coefficients). For pure mathematicians, this result may be great. But we want to really generate all the possibilities in a list.
 *)


let rec extract k ls =
  if k = 0 then [[]]
  else 
    let rec collect = function
      | [] -> []
      | x::xs -> List.map (fun y -> x::y) (extract (k-1) xs)::(collect xs)
    in
      collect ls |> List.concat

let _ = assert (
  List.sort Pervasives.compare (extract 2 ["a";"b";"c";"d"]) = 
  List.sort Pervasives.compare [["c"; "d"]; ["b"; "d"]; ["b"; "c"]; ["a"; "d"]; ["a"; "c"]; ["a"; "b"]])

File: prolog/21-30/26.pl
/*
Generate the combinations of K distinct objects chosen from the N elements of a list
    In how many ways can a committee of 3 be chosen from a group of 12 people? We all know that there 
    are C(12,3) = 220 possibilities (C(N,K) denotes the well-known binomial coefficients). For pure 
    mathematicians, this result may be great. But we want to really generate all the possibilities 
    (via backtracking).

    Example:
    ?- combination(3,[a,b,c,d,e,f],L).
    L = [a,b,c] ;
    L = [a,b,d] ;
    L = [a,b,e] ;
    ...
*/

collect(H, [H|T], T).
collect(X, [_|T], Cs) :- collect(X, T, Cs).

combination(0, _, []).
combination(K, Xs, [H|T]) :- 
  K > 0,
  collect(H, Xs, Zs),
  K1 is K - 1,
  combination(K1, Zs, T).


